tests/WE2E/monitor_jobs.py

```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/env python3
+
+import sys
+import argparse
+import logging
+import subprocess
+import sqlite3
+import time
+from textwrap import dedent
+from datetime import datetime
+from contextlib import closing
+
+sys.path.append("../../ush")
+
+from python_utils import (
+    load_config_file,
+    cfg_to_yaml_str
+)
+
+from check_python_version import check_python_version
+
+
+def monitor_jobs(expt_dict: dict, monitor_file: str = '', debug: bool = False) -> str:
+    """Function to monitor and run jobs for the specified experiment using Rocoto
+
+    Args:
+        expt_dict   (dict): A dictionary containing the information needed to run
+                            one or more experiments. See example file monitor_jobs.yaml
+        monitor_file (str): [optional]
+        debug       (bool): [optional] Enable extra output for debugging
+    Returns:
+        monitor_file (str): The name of the file used for job monitoring (when script
+                            is finish, this contains results/summary)
+
+    """
+
+    starttime = datetime.now()
+    # Write monitor_file, which will contain information on each monitored experiment
+    if not monitor_file:
+        monitor_file = f'monitor_jobs_{starttime.strftime("%Y%m%d%H%M%S")}.yaml'
+        logging.info(f"Writing information for all experiments to {monitor_file}")
+
+    write_monitor_file(monitor_file,expt_dict)
+
+    # Perform initial setup for each experiment
+    logging.info("Checking tests available for monitoring...")
+    num_expts = 0
+    for expt in expt_dict:
+        logging.info(f"Starting experiment {expt} running")
+        num_expts += 1
+        rocoto_db = f"{expt_dict[expt]['expt_dir']}/FV3LAM_wflow.db"
+        subprocess.run(["rocotorun", f"-w {expt_dict[expt]['expt_dir']}/FV3LAM_wflow.xml", f"-d {rocoto_db}"])
+        logging.debug(f"Reading database for experiment {expt}, populating experiment dictionary")
+        try:
+            db = sqlite_read(rocoto_db,'SELECT taskname,cycle,state from jobs')
+        except:
+            logging.warning(f"Unable to read database {rocoto_db}\nWill not track experiment {expt}")
+            expt_dict[expt]["status"] = "ERROR"
+            num_expts -= 1
+            continue
+        for task in db:
+            # For each entry from rocoto database, store that under a dictionary key named TASKNAME_CYCLE
+            # Cycle comes from the database in Unix Time (seconds), so convert to human-readable
+            cycle = datetime.utcfromtimestamp(task[1]).strftime('%Y%m%d%H%M')
+            expt_dict[expt][f"{task[0]}_{cycle}"] = task[2]
+        expt_dict[expt] = update_expt_status(expt_dict[expt], expt)
+        #Run rocotorun again to get around rocotobqserver proliferation issue
+        subprocess.run(["rocotorun", f"-w {expt_dict[expt]['expt_dir']}/FV3LAM_wflow.xml", f"-d {rocoto_db}"])
```

Comment:
Will the calls to subprocess like this wait until `rocotorun` completes before control is returned to the Python script? If not, you might be over-taxing the system here.

Also, is `rocotorun` also running in a cron job at this point?

Reply:
`subprocess.run` waits for the subprocess to complete before continuing.

The `monitor_jobs()` function is only called by `run_WE2E_tests.py` if the `--use_cron_to_relaunch` option is not specified, so this should never be run in conjunction with crontab running. But I can't control what users choose to do; theoretically this should work just fine if someone also had a crontab entry for this job.